pkg/network/usm/monitor.go

```diff
@@ -181,30 +180,45 @@ func (m *Monitor) Start() error {
 		return nil
 	}
 
-	var err error
+	var (
+		err error
+
+		// This value is here so that both the new way of handling protocols and
+		// the old one (used here by kafka & http2 monitoring) can coexist in
+		// this function. It SHOULD be removed once every protocol has been
+		// refactored to use the new way.
+		enabledCount int = 0
+	)
 
 	defer func() {
 		if err != nil {
 			if errors.Is(err, syscall.ENOMEM) {
 				err = fmt.Errorf("could not enable http monitoring: not enough memory to attach http ebpf socket filter. please consider raising the limit via sysctl -w net.core.optmem_max=<LIMIT>")
 			}
 
+			// Cleanup every remaining protocols
+			for _, protocol := range m.protocols {
+				protocol.PreStop(m.ebpfProgram.Manager.Manager)
+				protocol.PostStop(m.ebpfProgram.Manager.Manager)
+			}
+
 			if err != nil {
-				err = fmt.Errorf("could not enable http monitoring: %s", err)
+				err = fmt.Errorf("could not enable USM: %s", err)
 			}
 			startupError = err
 		}
 	}()
 
-	m.httpConsumer, err = events.NewConsumer(
-		"http",
-		m.ebpfProgram.Manager.Manager,
-		m.processHTTP,
-	)
-	if err != nil {
-		return err
+	for protocolType, protocol := range m.protocols {
+		err = protocol.PreStart(m.ebpfProgram.Manager.Manager)
+		if err != nil {
+			delete(m.protocols, protocolType)
+			log.Errorf("could not complete pre-start phase of %s monitoring: %s", protocolType, err)
+			continue
+		}
+
+		enabledCount++
```

Comment:
seems redundant
instead, after the loop:

```
enabledCount = len(m.protocols)
```

Reply:
So, I thought of doing that actually, but using `len(m.protocols)` will only work after all of the protocols are refactored. If we were to use it now, if, for example, http monitoring is disabled (by config or because of an error) but kafka is enabled, we would still abort the initialisation of USM because the length of `m.protocols` would be zero. Once the other protocols are refactored then yes, the `enabledCount` variables won't be needed anymore and we will able to just use `len(m.protocols)`.

Reply:
not sure I'm following the answer

Reply:
`m.enabledProtocols` contains instances of the `Protocol` interface. Currently, only http implements that interface and as such, only http can be in this map. So until we refactor the other protocols we support, `len(m.enabledProtocols)` can only be either 0 (if http is disabled) or 1 (if http is enabled). It won't reflect the actual number of enabled protocols until we refactor all of the protocols we support.